iosxr_iso2vbox.py

```diff
@@ -65,38 +63,295 @@
+        run(['VBoxManage', 'unregistervm', box_name, '--delete'])
 
-    In those cases the raw commands will need to be used.
 
-    E.g. passing grep 'a string separated by spaces'.
+def start_vboxheadless(args):
+    '''
+    Start vboxheadless process
     '''
-    argv = command_string.split()
-    if debug is True:
-        print ('argv: %s' % argv)
-    result = subprocess.check_output(argv)
-    return result
+    logger.debug('args: %s', args)
+    with open(os.devnull, 'w') as fp:
+        subprocess.Popen((args), stdout=fp)
+    time.sleep(2)
+
+
+def configure_xr(argv):
+    '''
+    Bring up XR and do some initial config.
+    Using socat to do the connection as telnet has an
+    odd double return on vbox
+    '''
+    logger.info('Logging into Vagrant Virtualbox and configuring IOS XR')
+
+    localhost = 'localhost'
+    prompt = "[$#]"
+
+    def xr_cli_wait_for_output(command, pattern):
+        '''
+        Execute a XR CLI command and try to find a pattern.
+        Try up to five times them register an error.
+        Each retry has a 5s turn around.
+        '''
+        total = 5
+        found_match = False
+
+        for attempt in range(total):
+            try:
+                child.sendline(command)
+                child.expect(pattern, 5)
+                found_match = True
+                break
+            except pexpect.TIMEOUT:
+                logger.debug("Iteration '%s' out of '%s'", attempt, total)
+
+        if not found_match:
+            raise Exception("No '%s' in '%s'" % (pattern, command))
+
+    try:
+        child = pexpect.spawn("socat TCP:%s:%s -,raw,echo=0,escape=0x1d" % (localhost, CONSOLE_PORT))
+
+        if argv == logging.DEBUG:
+            child.logfile = sys.stdout
+
+        child.timeout = 600  # Long time for full configuration, waiting for ip address etc
+
+        # Setup username and password and log in
+        child.expect('Press RETURN to get started.', child.timeout)
+        child.sendline("")  # Send enter
+        child.expect('Enter root-system username:', child.timeout)
+        child.sendline("vagrant")
+        child.expect("Enter secret:")
+        child.sendline("vagrant")
+        child.expect("Enter secret again:")
+        child.sendline("vagrant")
+        child.expect("Username:")
+        child.sendline("vagrant")
+        child.expect("Password:")
+        child.sendline("vagrant")
+        child.expect(prompt)
+
+        # Set term width
+        child.sendline("term width 300")
+        child.expect(prompt)
+
+        # ZTP causes some startup issues so disable it during box creation
+        child.sendline("ztp terminate noprompt")
+        child.expect(prompt)
+
+        # Determine if the image is a crypto/k9 image or not
+        # This will be used to determine whether to configure ssh or not
+        child.sendline("bash -c rpm -qa | grep k9sec")
+        output = child.expect('-k9sec')
```

Comment:
This isn't super-ideal as it means that for a non-k9 image, this will stall until expect() times out. Better to sendline and expect the next prompt (which should return quickly whether k9 or non-k9 regardless), then regexp against the received output to see if `-k9sec` is present.